### PR TITLE
Implement fertility & mortality ASR projection variants

### PR DESCRIPTION
**Pull request by LoraArmstrong**
In workflow, replace version 1.0.0. with version 1.1.0 for functions: project-asfr, project-asmr. Add required params & input data to be able to run these new functions in the workflow.

### CODE REVIEW

File: src/witan/models/dem/ccm/models.clj [hunk 1]
```diff
@@ -72,7 +73,10 @@
      :witan/version "1.0.0"
      :witan/type :function
      :witan/fn :ccm-core/add-births}
-    {:witan/name :age-on, :witan/version "1.0.0", :witan/type :function, :witan/fn :ccm-cor/age-on}
+    {:witan/name :age-on
+     :witan/version "1.0.0"
+     :witan/type :function
+     :witan/fn :ccm-cor/age-on}
```
Comment (acron0) [inline src/witan/models/dem/ccm/models.clj]: I notice there's a missing `e` in `core` - please check the contract and fix in either location, than ks.

File: src/witan/models/dem/ccm/models.clj [hunk 2]
```diff
@@ -167,14 +177,18 @@
...
      :witan/params {:start-year-avg-domout-mig 2003, :end-year-avg-domout-mig 2014}}
     {:witan/name :project-asfr
-     :witan/version "1.0.0"
+     :witan/version "1.1.0"
+     :witan/type :function
+     :witan/fn :ccm-fert/project-asfr
+     :witan/params {:fert-variant :fixed, :first-proj-year 2015, :last-proj-year 2018,
+                    :fert-scenario :principal-2012}}
+    {:witan/name :project-asmr
+     :witan/version "1.1.0"
```
Comment (acron0) [inline src/witan/models/dem/ccm/models.clj]: @sunnytownsend I remember @seb231 saying we weren't ready to give users the choice - are you sure we want to use variant-based functions for the `project-*` collection from now on?

File: src/witan/models/dem/ccm/inputs.clj [hunk 3]
```diff
@@ -44,6 +44,12 @@
    :witan/key :future-mortality-trend-assumption
    :witan/schema s/NationalTrendsSchema})
 
+(definput in-future-fert-trend-1-0-0
```
Comment (acron0) [inline src/witan/models/dem/ccm/inputs.clj:47]: I can't see a CSV in the default_data sets that seems to match this. Am I just blind?

File: src/witan/models/dem/ccm/fert/fertility.clj [hunk 4]
```diff
@@ -224,31 +231,13 @@
...
-   :witan/exported? true}
-  [{:keys [initial-projected-fertility-rates population-at-risk]} _]
-  (let [projected-births (cf/project-component-fixed-rates
-                          population-at-risk
-                          initial-projected-fertility-rates
-                          :fert-rate :births)]
-    {:births-by-age-sex-mother projected-births}))
-
-(defworkflowfn project-births-1-1-0
+(defworkflowfn project-births
```
Comment (acron0) [inline src/witan/models/dem/ccm/fert/fertility.clj]: I'd retain the clojureland function name (with version included)

File: src/witan/models/dem/ccm/models_utils.clj [hunk 5]
```diff
@@ -37,7 +37,7 @@
   [task-coll]
   (mapv (fn [[k v]]
           (let [m (hash-map :witan/name k
-                            :witan/version "1.0.0"
+                            :witan/version (:witan/version (get-meta v))
```
Comment (acron0) [inline src/witan/models/dem/ccm/models_utils.clj:40]: Good spot.

File: src/witan/models/dem/ccm/inputs.clj (hunk 3 above)
Comment (LoraArmstrong) [reply src/witan/models/dem/ccm/inputs.clj:47]: https://github.com/MastodonC/witan.models.demography/blob/feature/implement-fert-and-mort-variants/datasets/default_datasets/fertility/future_fertility_trend_assumption.csv

Comment (acron0) [reply src/witan/models/dem/ccm/inputs.clj:47]: Just blind!

File: src/witan/models/dem/ccm/fert/fertility.clj (hunk 4 above)
Comment (LoraArmstrong) [reply src/witan/models/dem/ccm/fert/fertility.clj]: even though this is now the only project-births function? did it this way to be consistent with project-deaths in mortality, but can change both of them if it's better to include version num.

File: src/witan/models/dem/ccm/models.clj (hunk 1 above)
Comment (LoraArmstrong) [reply src/witan/models/dem/ccm/models.clj]: well spotted...

File: src/witan/models/dem/ccm/fert/fertility.clj [hunk 6]
```diff
@@ -103,6 +103,15 @@
...
          (ds/select-columns [:gss-code :sex :age :year :fert-rate]))}))
 
+;;Age specific fertility rates projection
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Note: Fertility & mortality use different project ASR functions because future trends data ;;
+;; from ONS is different. Mortality future trend has a sex column whereas fertility future    ;;
+;; trend does not. Also, the future rates dataset needs to be aged on for mortality in the    ;;
+;; calc-proportional-differences function due to the way ages are used by the ONS, to give    ;;
+;; the correct death rates. This is not the case for the future fertility rates.              ;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/fert/fertility.clj:113]: 👍  to the clear documentation

File: src/witan/models/dem/ccm/inputs.clj (hunk 3 above)
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/inputs.clj:47]: why are there versions in the names in these inputs?

File: src/witan/models/dem/ccm/models.clj (hunk 1 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/models.clj]: Lora did the model run and tests pass despite this error? If so, feels like we are missing a test...

File: src/witan/models/dem/ccm/models.clj (hunk 2 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/models.clj]: Ultimately we need the variant based functions and the defaults to be set to `:applynationaltrend` for the project-asft variant (and `:average-applynationaltrend` for the project-asmr variant) to make the results be closer to ONS. Are there reasons why we can't do that in this PR @seb231 @LoraArmstrong?

File: src/witan/models/dem/ccm/models.clj [hunk 7]
```diff
@@ -225,19 +239,20 @@
...
        inputs/in-hist-total-births-1-0-0
        inputs/in-proj-births-by-age-of-mother-1-0-0
 
        ;; fertility fns
-       fert/project-asfr-1-0-0
-       fert/project-births-1-0-0
+       fert/project-asfr-1-1-0
+       fert/project-births
        fert/combine-into-births-by-sex
        fert/calculate-historic-asfr
```
Comment (sunnytownsend) [inline src/witan/models/dem/ccm/models.clj:250]: @acron0 should we do something about this inconsistency that some fns/inputs have a version in the name and some don't?

File: src/witan/models/dem/ccm/models.clj (hunk 2 above)
Comment (seb231) [reply src/witan/models/dem/ccm/models.clj]: I don't think so. Just like for asmr we set a default. Version 1.0.0 pointed to version 1.1.0. Version 1.0.0 has the open parameter set as a default, while 1.1.0 allows each variant

File: src/witan/models/dem/ccm/models.clj (hunk 1 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/models.clj]: was cor everywhere which is why it still ran, no need to add a test

File: src/witan/models/dem/ccm/models.clj (hunk 2 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/models.clj]: talked to @acron0 - agreed to get rid of version 1-0-0 and make version 1-1-0 to be the first version of this function now i.e.  1-0-0

File: src/witan/models/dem/ccm/fert/fertility.clj (hunk 4 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/fert/fertility.clj]: agreed with @acron0 put the version number back in - where we need to change these fns we should add version numbers

File: src/witan/models/dem/ccm/inputs.clj (hunk 3 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/inputs.clj:47]: @acron0 says because they are functions as well

File: src/witan/models/dem/ccm/models.clj (hunk 7 above)
Comment (sunnytownsend) [reply src/witan/models/dem/ccm/models.clj:250]: agreed with @acron0 not necessary to do it now, but if fns get edited, we should take the opportunity to add versions in the names. The reason for adding versions is to save having to describe how versions differ inside the name (i.e. avoiding things like project-asfr-average-applynationaltrend)

File: src/witan/models/dem/ccm/models.clj (hunk 1 above)
Comment (LoraArmstrong) [reply src/witan/models/dem/ccm/models.clj]: yes, it was everywhere. have changed it in both places now.

**Comment by LoraArmstrong** [comment]
@sunnytownsend, will add versions back in as requested. I can change the default variants to applynationaltrend, will have to see if it breaks any tests. The reason I set it to fixed was to match what was originally used in the R code, but if that's not what we want there's no reason to leave it like that.

**Comment by sunnytownsend** [comment]
:shipit:

**Comment by sunnytownsend** [comment]
apologies @LoraArmstrong - can you please change the 1-1-0 fns to be 1-0-0 now, so there is a resetting to all the fns being 1-0-0 for the first time users

**Comment by LoraArmstrong** [comment]
@sunnytownsend versions/names are back to 1-0-0 now

**Comment by sunnytownsend** [comment]
good job thanks @LoraArmstrong think its finally time to :shipit:

**Pull request merged**